File: Blackjack_Browser/scripts/table.js
class Table {

    deck = []; // Array of Cards
    currentCard = 0; // int
    numberOfDecks = 0; // int: number of decks at this table

    dealerHand = null; // Hand: dealer
    playerHand = null; // Hand: player // TODO: add multiple

    playerSplit = false; // boolean
    splitHands = []; // Array of Hands if player splits

    rules = null; // Rules

    constructor(numberOfDecks, rules=null) {
        this.numberOfDecks = numberOfDecks;
        this.rules = rules;
        this.initializeDeck();
        this.reset();
    }

    initializeDeck() {
        this.deck = [];
        this.currentCard = 0;

        let suits = ["CLUBS", "HEARTS", "SPADES", "DIAMONDS"];
        let values = ["ACE", "TWO", "THREE", "FOUR", "FIVE", "SIX", "SEVEN", "EIGHT", "NINE", "TEN", "JACK", "QUEEN", "KING"];

        for (let i = 0; i < this.numberOfDecks; i++) {
            for (let s = 0; s < 4; s++) {
                for (let v = 0; v < 13; v++) {
                    this.deck.push(new Card(Value[values[v]], Suit[suits[s]]));
                }
            }
        }
    }

    reset() {
        this.currentCard = 0;
        this.dealerHand = new Hand(name="Dealer");
        this.playerHand = new Hand(name="Player");
        this.playerSplit = false;
        this.splitHands = [];
        this.shuffleDeck();
    }

    shuffleDeck() {
        shuffle(this.deck);
    }

    getCard() {
        return this.deck[this.currentCard++];
    }

    deal() {
        this.playerHand.add(this.getCard());
        this.dealerHand.add(this.getCard());
        this.playerHand.add(this.getCard());
        this.dealerHand.add(this.getCard());
    }

    seeDealerCard() {
        if (this.dealerHand != null && this.dealerHand.lenght >= 2) {
            return this.dealerHand[1];
        } else {
            console.log("WARNING -- see delaer card fail");
            return null;
        }
    }

    // TODO: maybe we don't need this?
    hit(hand) {
        hand.add(this.getCard());
    }
    stand(hand) {
        hand.stand();
    }
    doubleDown(hand) {
        console.log(hand);
        hand.doubleDown(this.getCard());
    }
    split(hand) {
        this.playerSplit = true;
        let hand1 = new Hand();
        hand1.add(hand.cards[0]);
        hand1.enableSplit = false;
        let hand2 = new Hand();
        hand2.add(hand.cards[1]);
        hand2.enableSplit = false;

        hand1.add(this.getCard());
        hand2.add(this.getCard());

        hand1.betAmount = hand.betAmount;
        hand2.betAmount = hand.betAmount;

        this.splitHands.push(hand1);
        this.splitHands.push(hand2);
    }

    setBet(bet) {
        this.playerHand.betAmount = bet;
    }

    get playerHand() {
        return this.playerHand;
    }
    get dealerHand() {
        return this.dealerHand;
    }
    get splitHands() {
        return this.splitHands;
    }

}

File: src/si/casino/Table.java
package si.casino;

import si.casino.card.Card;
import si.casino.card.Suit;
import si.casino.card.Value;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Table {

    private List<Card> deck;
    private int currentCard;
    private int numberOfDecks;

    private Hand dealerHand;
    private Hand playerHand;

    private boolean playerSplit;
    private List<Hand> splitHands;

    private Rules rules;

    public Table(int numberOfDecks, Rules rules) {
        // TODO: merge numOfDecks into rules
        this.numberOfDecks = numberOfDecks;
        this.rules = rules;
        initializeDeck();
        reset();
    }

    private void initializeDeck() {
        deck = new ArrayList<>();
        this.currentCard = 0;

        for (int i = 0; i < this.numberOfDecks; i++) {
            for (int s = 0; s < 4; s++) {
                for (int v = 0; v < 13; v++) {
                    this.deck.add(new Card(Value.values()[v], Suit.values()[s]));
                }
            }
        }
    }

    private void shuffleDeck() {
        Collections.shuffle(deck);
    }

    public void reset() {
        this.currentCard = 0;
        this.dealerHand = new Hand(rules);
        this.playerHand = new Hand(rules);
        this.playerSplit = false;
        this.splitHands = new ArrayList<>();
        shuffleDeck();
    }

    public Card getCard() {
        return deck.get(this.currentCard++);
    }

    public void deal() {
        playerHand.add(getCard());
        dealerHand.add(getCard());
        playerHand.add(getCard());
        dealerHand.add(getCard());
    }

    public Card seeDealerCard() {
        if (this.dealerHand != null && this.dealerHand.getSize() >= 2) {
            return dealerHand.getCards().get(1);
        } else {
            return null;
        }
    }

    public Hand getPlayerHand() {
        return this.playerHand;
    }

    public Hand getDealerHand() {
        return this.dealerHand;
    }

    public void hit(Hand hand) {
        hand.add(this.getCard());
    }

    public void stand(Hand hand) {
        hand.stand();
    }

    public void doubleDown(Hand hand) {
        hand.doubleDown(this.getCard());
    }

    public void split(Hand hand) {

        // TODO: support for more than 1 split?

        this.playerSplit = true;
        Hand hand1 = new Hand(rules);
        hand1.add(hand.getCards().get(0));
        hand1.setEnableSplit(false);
        Hand hand2 = new Hand(rules);
        hand2.add(hand.getCards().get(1));
        hand2.setEnableSplit(false);

        hand1.add(getCard());
        hand2.add(getCard());

        hand1.setBetAmount(hand.getBetAmount());
        hand2.setBetAmount(hand.getBetAmount());

        this.splitHands.add(hand1);
        this.splitHands.add(hand2);
    }

    public void setBet(int bet) {
        this.playerHand.setBetAmount(bet);
    }

    public boolean isPlayerSplit() {
        return playerSplit;
    }

    public List<Hand> getSplitHands() {
        return splitHands;
    }

}
